Create scala_proto_library rule using scalaPB

azymnis opened the pull request:
This addresses issue #144

@johnynek @ittaiz I put this up to get your feedback. This currently "works" but I need to test it more rigorously, add all available options that can be passed to scalapb, etc.

Note that I'm taking a different approach to the scrooge rules: there's only one parameter that you can pass into the rule and that is `deps`, which has to be of type `proto_library`. Since `proto_library` already does the work of figuring out all transitive proto file dependencies, there is a lot less work that this rule has to do. This would not allow for remote proto jars, but I would argue that using these is a really bad idea in the first place.

Anyway, let me know what you think.

bazel-io commented:
Can one of the admins verify this patch?

ittaiz commented:
test this please (ci bot)

azymnis commented:
@ittaiz the build seems to have failed. Is this a macos issue?

johnynek commented:
yeah, passes on linux. The bazel ci is broken (needs to accept xcode license) cc @damienmg 

The travis macos timed out. Maybe a transient travis issue.

Will review!

Thanks man, this is great!

johnynek commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,169 @@
+    )
+
+    native.maven_jar(
+        name = "scala_proto_rules_slalapbc",
```
looks like a typo here: `slala`

azymnis replied:
Whoops let me fix

johnynek commented on src/scala/scripts/ScalaPBGenerator.scala:
```diff
@@ -0,0 +1,68 @@
+    val jarOutput = args.get(0)
+    val protoFiles = args.get(1).split(':').toList
+
+    val tmp = Paths.get(Option(System.getenv("TMPDIR")).getOrElse("/tmp"))
```
I think we actually want:

`System.getProperty("java.io.tmpdir")`

Which I think bazel sets up correctly.

we got burned on this somewhere else in some context. Did you copy this from another rule?

azymnis replied:
Yeah I copied this from the scrooge rules. Want me to fix that one too?

johnynek commented on test/BUILD:
```diff
@@ -432,3 +435,8 @@ scala_binary(
+scala_proto_library(
+    name = "test_proto",
+    deps = ["//test/proto:test2", "//test/proto:test3"]
+)
```
can you add a test that creates these things in some scala code?

azymnis replied:
Sure, let me do that... I'll also add a mock grpc service to see how that works too

johnynek commented:
Yeah, if you could fix the Scrooge rules it would be great.

On Tue, Sep 26, 2017 at 18:31 Argyris Zymnis <notifications@github.com>
wrote:

> *@azymnis* commented on this pull request.
> ------------------------------
>
> In test/BUILD
> <https://github.com/bazelbuild/rules_scala/pull/281#discussion_r141242534>
> :
>
> > @@ -432,3 +435,8 @@ scala_binary(
>      main_class = "scala.test.BinaryDependentOnJava",
>      deps = [":JavaOnlySources"],
>  )
> +
> +scala_proto_library(
> +    name = "test_proto",
> +    deps = ["//test/proto:test2", "//test/proto:test3"]
> +)
>
> Sure, let me do that... I'll also add a mock grpc service to see how that
> works too
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/281#discussion_r141242534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEJdoy4nhL1FxQ3A6Ku_VA0hMNQq5hBks5smc-fgaJpZM4PktRQ>
> .
>
-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin

azymnis commented:
@johnynek so I went ahead and added support for grpc service definitions but this opens up a whole can of worms: grpc depends on a ton of dependencies that I had to pull in. Added a flag to turn grpc service stub generation on and off for a target (this is supported by scalapb compiler). 

I also added a test for a dumb grpc server and client. The code compiles and runs on my machine, but not sure what the easiest way would be to actually run this in CI.

Anyway, let me know what you think. Not happy that we need to pull in the universe to get grpc bindings to work, but not sure that there's really anything that we can do about it.

damienmg commented:
Jenkins: retest this please

(license should be fixed on all the mac slave now)

azymnis commented:
@damienmg there was an error in one of my BUILD files, which I just fixed

@johnynek @ittaiz what do you think of calling the rule `scalapb_proto_library`?

johnynek commented:
test this please

johnynek commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,407 @@
+    "//external:io_bazel_rules_scala/dependency/proto/scalapb_fastparse",
+]
+
+GRPC_DEPS = [
```
deep dependency stack. :/

azymnis replied:
Yeah it's kind of nuts actually

johnynek commented:
This looks good to me.

👍

azymnis commented:
I just added all other available scalapbc flags to this (`java_conversions`, `flat_package`, `single_line_to_string`).

BTW I noticed that some of the travis builds time out. Is that a regression that I introduced or has it been going on for a while? This build is really really slow. Should we maybe break it up into separate jobs? Is calling `bazel clean` the cause of slowness?

johnynek commented:
test this please

johnynek commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,442 @@
+        executable = ctx.executable._pluck_scalapb_scala,
+        inputs = list(acc_imports) + [argfile],
+        outputs = [ctx.outputs.srcjar],
+        mnemonic="ProtoScalaRule",
```
should this be `ProtoScalaPbRule` or something?

azymnis replied:
Yeah let me rename it

azymnis commented:
Any ideas about what to do with build timeouts?

johnynek commented:
test this please

azymnis commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,442 @@
+        "with_grpc": attr.bool(default=False),
+        "with_java_conversions": attr.bool(default=False),
+        "with_flat_package": attr.bool(default=False),
+        "with_single_line_to_string": attr.bool(default=False),
```
This makes scalapb generate a `toString` method for generated classes that prints a single line of text. Not sure why you would want that but it is available as a compiler option.

azymnis commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,442 @@
+        ),
+        "with_grpc": attr.bool(default=False),
+        "with_java_conversions": attr.bool(default=False),
+        "with_flat_package": attr.bool(default=False),
```
Here's where all the scalapb compiler settings are explained https://scalapb.github.io/sbt-settings.html

I'm happy to add comments for all of these

johnynek replied:
yeah, adding a link to these settings in the comment would be good.

azymnis commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,442 @@
+            mandatory=True,
+            allow_rules=["proto_library"]
+        ),
+        "with_grpc": attr.bool(default=False),
```
Will do

azymnis commented on scala_proto/scala_proto.bzl:
```diff
@@ -0,0 +1,442 @@
+
+    external_deps = list(SCALAPB_DEPS + GRPC_DEPS if (with_grpc) else SCALAPB_DEPS)
+
+    if with_java_conversions:
```
Sure. Maybe I should just remove all the `with_` prefixes?

johnynek replied:
sounds good.

azymnis replied:
Actually, sorry misread your comment. Sure let me just change that

azymnis commented:
@johnynek I merged your branch and this seems to have made the @googlebot unhappy

johnynek commented:
it's okay, that bot's not the boss of me.

johnynek commented:
this is no longer WIP, right?

Ready to merge @azymnis ?

azymnis commented:
Yup ready!

On Fri, Sep 29, 2017 at 5:12 PM P. Oscar Boykin <notifications@github.com>
wrote:

> this is no longer WIP, right?
>
> Ready to merge @azymnis <https://github.com/azymnis> ?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/281#issuecomment-333266526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMpF-Yy86gNlQByl3mIwbSS7H8HSvViks5snYdqgaJpZM4PktRQ>
> .
>

johnynek commented:
👍 

I think I'll let @ittaiz merge #285 first (not squash/rebase I hope) then I will merge this after that.

azymnis commented:
@johnynek I should prob update the rule docs too in the readme. Hold off before merging till I do that

johnynek commented:
@azymnis thanks so much for doing this!

I'll tweet about this on Monday (to maximize exposure). I think the fact that the bazel rules now support proto and grpc is a big selling point. I hope this gets us more users, bringing more testing and more contributors!

Thanks again!

azymnis commented:
@johnynek no problem, happy to work on this and learn more about bazel

ittaiz commented:
Indeed thank you for this valuable addition.
I was out of the loop due to holidays and backlog but as we (Wix) are moving a lot of our RPC to gRpc this will be heavily used.

ittaiz commented on src/scala/scripts/BUILD:
```diff
@@ -21,3 +21,29 @@ scala_binary(
+        "//src/java/io/bazel/rulesscala/io_utils",
+        "//src/java/io/bazel/rulesscala/jar",
+        "//src/java/io/bazel/rulesscala/worker",
+        "//src/scala/io/bazel/rules_scala/scrooge_support:compiler",
```
Why does protobuf generator need to depend on scrooge?

azymnis replied:
Fix is here: https://github.com/bazelbuild/rules_scala/pull/290/files

ittaiz commented on src/scala/scripts/ScalaPBGenerator.scala:
```diff
@@ -0,0 +1,72 @@
+    val scalaPBArgs = s"--scala_out=${flagPrefix}${scalaPBOutput}" :: protoFiles
+    val config = ScalaPBC.processArgs(scalaPBArgs.toArray)
+    val code = ProtocBridge.runWithGenerators(
+      protoc = a => com.github.os72.protocjar.Protoc.runProtoc(config.version +: a.toArray),
```
This library jar brings with it the compilers for the relevant platforms right? It doesn't download anything during build time according to the running platform, right?

azymnis replied:
Yes that's right. My understanding while reading the ScalaPBC doc is that it depends on `protoc-bridge` and `protobuf-java` which I have added as explicit dependencies.

azymnis commented:
Oh good point, it shouldn't depend on Scrooge. I'll send a fix

On Sat, Sep 30, 2017 at 10:58 PM Ittai Zeidman <notifications@github.com>
wrote:

> *@ittaiz* commented on this pull request.
> ------------------------------
>
> In src/scala/scripts/BUILD
> <https://github.com/bazelbuild/rules_scala/pull/281#discussion_r142021428>
> :
>
> > @@ -21,3 +21,29 @@ scala_binary(
>          ":generator_lib",
>      ],
>  )
> +
> +scala_library(
> +    name = "scalapb_generator_lib",
> +    srcs = ["ScalaPBGenerator.scala"],
> +    visibility = ["//visibility:public"],
> +    deps = [
> +        "//external:io_bazel_rules_scala/dependency/proto/protoc",
> +        "//external:io_bazel_rules_scala/dependency/proto/scalapb_plugin",
> +        "//external:io_bazel_rules_scala/dependency/proto/protoc_bridge",
> +        "//external:io_bazel_rules_scala/dependency/proto/scalapbc",
> +        "//external:io_bazel_rules_scala/dependency/com_google_protobuf/protobuf_java",
> +        "//src/java/io/bazel/rulesscala/io_utils",
> +        "//src/java/io/bazel/rulesscala/jar",
> +        "//src/java/io/bazel/rulesscala/worker",
> +        "//src/scala/io/bazel/rules_scala/scrooge_support:compiler",
>
> Why does protobuf generator need to depend on scrooge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/281#pullrequestreview-66328567>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMpFzdDzHUSKj49JXBhBLAa8ieyNgpcks5snyoWgaJpZM4PktRQ>
> .
>
